crates/uv/src/commands/project/add.rs

```diff
@@ -28,10 +50,101 @@ pub(crate) async fn add(
     // Find the project requirements.
     let project = ProjectWorkspace::discover(&std::env::current_dir()?, None).await?;
 
+    // Discover or create the virtual environment.
+    let venv = project::init_environment(
+        project.workspace(),
+        python.as_deref(),
+        preview,
+        cache,
+        printer,
+    )?;
+
+    let client_builder = BaseClientBuilder::new()
+        .connectivity(connectivity)
+        .native_tls(native_tls)
+        .keyring(keyring_provider.unwrap_or_default());
+
+    let requirements = requirements
+        .into_iter()
+        .map(RequirementsSource::Package)
+        .collect_vec();
+
+    let RequirementsSpecification { requirements, .. } =
+        RequirementsSpecification::from_sources(&requirements, &[], &[], &client_builder).await?;
+
+    let hasher = HashStrategy::None;
+    let top_level_index = InMemoryIndex::default();
+    // We only need to resolve unnamed requirements, so this doesn't really matter.
+    let concurrency = Concurrency::default();
+
+    // Determine the environment for the resolution.
+    let (tags, markers) =
+        resolution_environment(python_version, python_platform, venv.interpreter())?;
+
+    // Initialize the registry client.
+    let client = RegistryClientBuilder::new(cache.clone())
+        .native_tls(native_tls)
+        .connectivity(connectivity)
+        .index_urls(index_locations.index_urls())
+        .index_strategy(index_strategy)
+        .keyring(keyring_provider.unwrap_or_default())
+        .markers(&markers)
+        .platform(venv.interpreter().platform())
+        .build();
+
+    // Resolve the flat indexes from `--find-links`.
+    let flat_index = {
+        let client = FlatIndexClient::new(&client, cache);
+        let entries = client.fetch(index_locations.flat_index()).await?;
+        FlatIndex::from_entries(entries, Some(&tags), &hasher, &no_build, &NoBinary::None)
```

Comment:
Can we define this `NoBinary` as `let no_binary = NoBinary::default()`, and put it with all the other defaults below? Like, move that block up to line 75 or so?

Reply:
I'd like to have something that looks like the `// TODO(charlie): Respect project configuration.` block in `do_lock`, just so we have consistency and it's easier to fix them all later.